src/node/historical_queries.h

```diff
@@ -0,0 +1,341 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "consensus/ledger_enclave_types.h"
+#include "kv/store.h"
+#include "node/historical_queries_interface.h"
+#include "node/history.h"
+#include "node/rpc/node_interface.h"
+
+#include <deque>
+#include <map>
+#include <memory>
+
+namespace ccf::historical
+{
+  class StateCache : public AbstractStateCache
+  {
+  protected:
+    kv::Store& source_store;
+    ringbuffer::WriterPtr to_host;
+
+    enum class RequestStage
+    {
+      Fetching,
+      Untrusted,
+      Trusted,
+    };
+
+    using LedgerEntry = std::vector<uint8_t>;
+
+    struct Request
+    {
+      RequestStage current_stage = RequestStage::Fetching;
+      crypto::Sha256Hash entry_hash = {};
+      StorePtr store = nullptr;
+    };
+
+    // TODO: Make this an LRU, evict old entries
```

Comment:
Question: Is LRU the right thing to do here? Consider doing something time based, it might be better to be slow rather than fail to return results if (MAX_ACTIVE_REQUESTS+1) clients are constantly trying perform historical queries.

Reply:
This is a good option. I think the LRU makes it easier to reason about the memory useage (to enforce a cap, we cull old requests), but we could instead say "remove entries after a day". Might be more general.

Our current use case intends to limit historical queries anyway (eg - an ACL so only some users can call it, assume coordination between users). Since they offer control over node memory useage/performance, I think they should be restricted in general. But that doesn't imply global coordination, so maybe the timeout is preferable. Will give this some more thought!